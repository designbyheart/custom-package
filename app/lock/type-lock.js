// @flow
// TODO:KS Move this type to common app types
import type {
  ReactNavigation,
} from '../qr-connection-request/type-qr-connection-request'
export const LOCK_TYPE = {
  FINGERPRINT: 'FINGERPRINT',
  PIN: 'PIN',
}

export type lockType = $Keys<typeof LOCK_TYPE>

export const PENDING_REDIRECT = 'SET_PENDING_REDIRECT'
export type AddPendingRedirectAction = {
  type: typeof PENDING_REDIRECT,
  routeName: string,
}

export const CLEAR_PENDING_REDIRECT = 'CLEAR_PENDING_REDIRECT'
export type ClearPendingRedirectAction = {
  type: typeof CLEAR_PENDING_REDIRECT,
}

export const PIN_STORAGE_KEY = 'APP_PIN_LOCK'
export const SET_PIN = 'SET_PIN'
export type SetPinAction = {
  type: typeof SET_PIN,
  pin: string,
}

export const CHECK_PIN = 'CHECK_PIN'
export type CheckPinAction = {
  type: typeof CHECK_PIN,
  pin: string,
}

export const CHECK_PIN_SUCCESS = 'CHECK_PIN_SUCCESS'
export type CheckPinSuccessAction = {
  type: typeof CHECK_PIN_SUCCESS,
}

export const CHECK_PIN_FAIL = 'CHECK_PIN_FAIL'
export type CheckPinFailAction = {
  type: typeof CHECK_PIN_FAIL,
}

export const CHECK_PIN_IDLE = 'CHECK_PIN_IDLE'

export const UNLOCK_APP = 'UNLOCK_APP'
export type UnlockAppAction = {
  type: typeof UNLOCK_APP,
}

export type CheckPinStatus =
  | typeof CHECK_PIN_SUCCESS
  | typeof CHECK_PIN_FAIL
  | typeof CHECK_PIN_IDLE

export type LockStore = {
  pendingRedirection: ?string,
  checkPinStatus: CheckPinStatus,
  isAppLocked: boolean,
}

export type LockActions =
  | AddPendingRedirectAction
  | ClearPendingRedirectAction
  | SetPinAction
  | CheckPinAction
  | CheckPinSuccessAction
  | CheckPinFailAction

export type LockEnterPinProps = {
  checkPinAction: (pin: string) => CheckPinAction,
  checkPinStatus: CheckPinStatus,
  pendingRedirection: string,
} & ReactNavigation
