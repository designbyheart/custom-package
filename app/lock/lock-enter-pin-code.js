// @flow
import React, { PureComponent } from 'react'
import { InteractionManager, StyleSheet } from 'react-native'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { checkPinAction, checkPinStatusIdle } from './lock-store'
import {
  Container,
  CustomText,
  CustomButton,
  PinCodeBox,
  CustomView,
} from '../components'
import { CHECK_PIN_IDLE, CHECK_PIN_SUCCESS, CHECK_PIN_FAIL } from './type-lock'
import type { Store } from '../store/type-store'
import type { LockEnterPinProps, LockEnterPinState } from './type-lock'
import { switchErrorAlerts } from '../store'

export const TitleText = (
  <CustomText h4 center bold>
    Enter pin code to unlock app
  </CustomText>
)
export const WrongPinText = (
  <CustomText h4 center>
    Wrong pin! Please try again
  </CustomText>
)

export class LockEnterPin extends PureComponent<
  void,
  LockEnterPinProps,
  LockEnterPinState
> {
  state = {
    interactionsDone: false,
  }

  pinCodeBox = null

  onPinComplete = (pin: string) => {
    // user entered 6 digits in pin box
    this.props.checkPinAction(pin)
  }

  clearFailStatus = () => {
    this.props.checkPinStatusIdle()
  }

  clearFailStatusDelayed = () => {
    setTimeout(this.clearFailStatus, 1000)
  }

  componentWillReceiveProps(nextProps: LockEnterPinProps) {
    if (this.props.checkPinStatus !== nextProps.checkPinStatus) {
      if (nextProps.checkPinStatus === CHECK_PIN_SUCCESS) {
        if (this.props.pendingRedirection) {
          this.props.navigation.navigate(this.props.pendingRedirection)
        }
      } else if (nextProps.checkPinStatus === CHECK_PIN_FAIL) {
        this.pinCodeBox && this.pinCodeBox.clear()
        // set status back to idle so we can come to this else again
        this.clearFailStatusDelayed()
      }
    }
  }

  componentDidMount() {
    InteractionManager.runAfterInteractions(() => {
      this.setState({ interactionsDone: true })
    })
  }

  render() {
    const { checkPinStatus } = this.props
    return (
      <Container primary>
        <CustomView
          style={[styles.text]}
          onPress={this.props.switchErrorAlerts}
        >
          {TitleText}
          {checkPinStatus === CHECK_PIN_FAIL && WrongPinText}
        </CustomView>
        <CustomView center>
          {this.state.interactionsDone && (
            <PinCodeBox
              ref={pinCodeBox => {
                this.pinCodeBox = pinCodeBox
              }}
              onPinComplete={this.onPinComplete}
            />
          )}
        </CustomView>
      </Container>
    )
  }
}

const mapStateToProps = (state: Store) => ({
  checkPinStatus: state.lock.checkPinStatus,
  pendingRedirection: state.lock.pendingRedirection,
})

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      checkPinAction,
      checkPinStatusIdle,
      switchErrorAlerts,
    },
    dispatch
  )

export default connect(mapStateToProps, mapDispatchToProps)(LockEnterPin)

const styles = StyleSheet.create({
  text: {
    marginTop: 120,
    marginBottom: 70,
  },
})
