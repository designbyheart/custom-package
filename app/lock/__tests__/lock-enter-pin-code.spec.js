// @flow
import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'
import { CHECK_PIN_IDLE } from '../type-lock'
import type { LockEnterPinProps } from '../type-lock'
import { LockEnterPin } from '../lock-enter-pin-code'

describe('<LockPinCodeEnter />', () => {
  // TODO:KS Figure out why this test is failing
  xit('should render pin code box', () => {
    const props: LockEnterPinProps = {
      checkPinAction: jest.fn(),
      checkPinStatusIdle: jest.fn(),
      checkPinStatus: CHECK_PIN_IDLE,
      pendingRedirection: '',
      navigation: {
        navigate: jest.fn(),
      },
    }
    const wrapper = renderer.create(<LockEnterPin {...props} />).toJSON()
    expect(wrapper).toMatchSnapshot()
  })

  // TODO:KS Complete these tests
  xit('should show wrong pin, if pin does not match', () => {})
  xit('should redirect to pendingRedirection after pin is success', () => {})
})
