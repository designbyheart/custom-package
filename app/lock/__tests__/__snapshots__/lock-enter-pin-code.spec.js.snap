// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LockPinCodeEnter /> should render pin code box 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f0f0f0",
      "flex": 1,
    }
  }
  tertiary={true}
>
  <BVLinearGradient
    colors={
      Array [
        4293585642,
        15790320,
      ]
    }
    end={undefined}
    locations={
      Array [
        0.08,
        1,
      ]
    }
    start={undefined}
    style={
      Array [
        Object {
          "height": 10,
        },
      ]
    }
  />
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    hitSlop={undefined}
    onLayout={undefined}
    onPress={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "marginBottom": 70,
        "marginTop": 120,
      }
    }
    testID={undefined}
  />
  <View
    center={true}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={Object {}}
    >
      <View
        row={true}
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          center={true}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#777777",
              "borderBottomWidth": 5,
              "height": 66,
              "justifyContent": "center",
              "marginHorizontal": 5,
              "width": 38,
            }
          }
          testID="pin-code-digit-0"
        />
        <View
          center={true}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#777777",
              "borderBottomWidth": 5,
              "height": 66,
              "justifyContent": "center",
              "marginHorizontal": 5,
              "width": 38,
            }
          }
          testID="pin-code-digit-1"
        />
        <View
          center={true}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#777777",
              "borderBottomWidth": 5,
              "height": 66,
              "justifyContent": "center",
              "marginHorizontal": 5,
              "width": 38,
            }
          }
          testID="pin-code-digit-2"
        />
        <View
          center={true}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#777777",
              "borderBottomWidth": 5,
              "height": 66,
              "justifyContent": "center",
              "marginHorizontal": 5,
              "width": 38,
            }
          }
          testID="pin-code-digit-3"
        />
        <View
          center={true}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#777777",
              "borderBottomWidth": 5,
              "height": 66,
              "justifyContent": "center",
              "marginHorizontal": 5,
              "width": 38,
            }
          }
          testID="pin-code-digit-4"
        />
        <View
          center={true}
          style={
            Object {
              "alignItems": "center",
              "borderBottomColor": "#777777",
              "borderBottomWidth": 5,
              "height": 66,
              "justifyContent": "center",
              "marginHorizontal": 5,
              "width": 38,
            }
          }
          testID="pin-code-digit-5"
        />
      </View>
      <TextInput
        autoCorrect={false}
        autoFocus={true}
        blurOnSubmit={false}
        enablesReturnKeyAutomatically={false}
        keyboardAppearance="dark"
        keyboardType="number-pad"
        maxLength={6}
        onChangeText={[Function]}
        style={
          Object {
            "position": "absolute",
            "right": -999,
          }
        }
        testID="pin-code-input-box"
        value=""
      />
    </View>
  </View>
</View>
`;
